akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala

```diff
@@ -867,6 +881,92 @@ private[akka] object Running {
+    setup.setMdcPhase(PersistenceMdc.WaitingAsyncEffect)
+    recursiveUnstashOne = 0
+    setup.context.pipeToSelf(laterEffect) {
+      case Success(effect) => AsyncEffectCompleted(msg, effect)
+      case Failure(exc)    => AsyncEffectCompleted(msg, Effect.none[E, S].thenRun(_ => throw exc))
+    }
+    new WaitingAsyncEffect(state)
+  }
+
+  /** INTERNAL API */
+  @InternalApi private[akka] class WaitingAsyncEffect(var state: RunningState[S])
+      extends AbstractBehavior[InternalProtocol](setup.context)
+      with WithSeqNrAccessible {
+
+    override def onMessage(msg: InternalProtocol): Behavior[InternalProtocol] = {
+      msg match {
+        case completed: AsyncEffectCompleted[C @unchecked, E @unchecked, S @unchecked] =>
+          onAsyncEffectCompleted(completed.cmd, completed.effect)
+        case in: IncomingCommand[C @unchecked]         => onCommand(in)
+        case re: ReplicatedEventEnvelope[E @unchecked] => onReplicatedEvent(re)
+        case pe: PublishedEventImpl                    => onPublishedEvent(pe)
+        case JournalResponse(r)                        => onDeleteEventsJournalResponse(r, state.state)
+        case SnapshotterResponse(r)                    => onDeleteSnapshotResponse(r, state.state)
+        case get: GetState[S @unchecked]               => stashInternal(get)
+        case getSeqNr: GetSeenSequenceNr               => onGetSeenSequenceNr(getSeqNr)
+        case _                                         => Behaviors.unhandled
+      }
+    }
+
+    def onAsyncEffectCompleted(cmd: C, effect: Effect[E, S]): Behavior[InternalProtocol] = {
+      val handlingCommands = new HandlingCommands(state)
+      handlingCommands.runEffect(cmd, effect)
+    }
+
+    def onCommand(cmd: IncomingCommand[C]): Behavior[InternalProtocol] = {
+      if (state.receivedPoisonPill) {
+        if (setup.settings.logOnStashing)
+          setup.internalLogger.debug("Discarding message [{}], because actor is to be stopped.", cmd)
+        Behaviors.unhandled
+      } else {
+        stashInternal(cmd)
+      }
+    }
+
+    def onGetSeenSequenceNr(get: GetSeenSequenceNr): WaitingAsyncEffect = {
+      get.replyTo ! state.seenPerReplica.getOrElse(get.replica, 0L)
+      this
+    }
+
+    def onReplicatedEvent(event: InternalProtocol.ReplicatedEventEnvelope[E]): Behavior[InternalProtocol] = {
+      if (state.receivedPoisonPill) {
+        Behaviors.unhandled
+      } else {
+        stashInternal(event)
+      }
+    }
+
+    def onPublishedEvent(event: PublishedEventImpl): Behavior[InternalProtocol] = {
+      if (state.receivedPoisonPill) {
+        Behaviors.unhandled
+      } else {
+        stashInternal(event)
+      }
+    }
+
+    override def onSignal: PartialFunction[Signal, Behavior[InternalProtocol]] = {
+      case PoisonPill =>
+        // wait for completion of async effect before stopping
+        state = state.copy(receivedPoisonPill = true)
```

Comment:
Hmm, is that really the right thing to do, poison pill will be abrupt termination kinda scenarios where passivation timed out etc. I think waiting for an arbitrary long future is different from PoisonPill when waiting for a write to complete (which is more or less time bounded)

Reply:
> poison pill will be abrupt termination kinda scenarios

I'm not sure I agree with that. PoisonPill is like a application specific stop message, but we have built-in support for it. E.g. it can be used as stop message in cluster sharding. The promise is that it will be handled after other messages in the mailbox. Here, such other messages can be stashed while waiting for persist or waiting for async effect. I think it's more consistent to wait for async effect in the same way as for other stashed commands.

Abrupt termination is done with `stop`, which cluster sharding will use if entities are not terminated within a timeout.

Reply:
Alright, thanks for clarifying.